lib/chunck/irmin_chunck.ml

```diff
@@ -0,0 +1,372 @@
+	    Chunck.extract_data x 0 result 0 dlen;
+	    Lwt.return (Some result)
+	 | Chunck.Indirect ->
+	    (Tree.walk_to_list t x) >>=
+	      (fun y -> 
+	       let result = Cstruct.concat y in
+	       return (Some result)
+	      )
+	)
+
+	  
+    let read_exn t key =
+      AO.read_exn t.db key >>=
+	(fun x ->
+	 match Chunck.type_of_chunck x with
+	 | Chunck.Data ->
+	    let dlen = Chunck.get_length x in
+	    let result = Cstruct.create dlen in
+	    Chunck.extract_data x 0 result 0 dlen;
+	    return result
+	 | Chunck.Indirect ->
+	    (Tree.walk_to_list t x) >>=
+	      (fun y ->
+	       let result = Cstruct.concat y in
+	       return result
+	      ))
+
+	  
+    let add t v =
+      let value_length = Cstruct.len v in
+      let rest_value_length = value_length mod t.data_length in
+
+      if value_length <= t.data_length then
+	let chunck = Chunck.create t value_length Chunck.Data in
+	Chunck.set_data v 0 chunck 0 value_length;  
+	AO.add t.db chunck >>= (fun x -> Lwt.return x)
+
+      else
+	let nbr_data_bloc, last_data_bloc_length =
+	let x =  value_length / t.data_length in
+	if rest_value_length == 0 then
+	  x, t.data_length
+	else
+	  (x + 1), rest_value_length
+      in
+      let nbr_indir_bloc, last_indir_bloc_length =
+	let x = nbr_data_bloc / t.nb_indirect in
+	let y = (nbr_data_bloc mod t.nb_indirect) in
+	if y == 0 then
+	  x, t.nb_indirect
+	else
+	  (x + 1), y 
+      in
+      
+      let split_data_first_level offset_second_level =
+	
+	let number_chunck =
+	  if offset_second_level == (nbr_indir_bloc - 1) then
+	    last_indir_bloc_length
+	  else
+	    t.nb_indirect
+	in
+	
+	let indir = Chunck.create t number_chunck Chunck.Indirect in
+	
+	let rec loop offset_first_level =
+	  let dlen, max_loop =
+	    if (offset_second_level == (nbr_indir_bloc - 1)) then
+	      if (offset_first_level == (last_indir_bloc_length - 1)) then
+		last_data_bloc_length, last_indir_bloc_length
+	      else
+		t.data_length, last_indir_bloc_length 
+	    else
+	      t.data_length, t.nb_indirect 
+	  in
+	  
+	  if offset_first_level >= max_loop then
+	    Lwt.return_unit
+	  else
+	    let chunck = Chunck.create t dlen Chunck.Data in
+	    let offset_hash = offset_first_level * t.hash_length in
+	    let offset_value = (offset_second_level * t.nb_indirect * t.data_length) + (offset_first_level * t.data_length) in 	    
+	    Chunck.set_data v offset_value chunck 0 dlen;  
+	    let add_chunck () =
+	      AO.add t.db chunck >>= fun x ->
+	      Chunck.set_data (K.to_raw x) 0 indir offset_hash Chunck.hash_length;
+	      Lwt.return_unit
+	    in
+	    Lwt.join [add_chunck (); loop (offset_first_level + 1)]
+	in
+	loop 0 >>= (fun _ ->
+		    AO.add t.db indir >>=
+		      (fun x ->
+		       Lwt.return x)
+		   )
+      in
+      
+      let main_loop () =
+	let rec first_level i =
+	  match i with
+	  | e when e == nbr_indir_bloc -> 
+	      Lwt.return []
+	  | e when e < nbr_indir_bloc ->
+	     (split_data_first_level i) >>=
```

Comment:
you can remove almost all the parenthesis in that block. The only needed one are the ones around `(i+1)` and `(x::y)`.

Also you can avoid the somehow verbose `>>= fun y -> Lwt.return (x::y)` and use `>|= fun y -> x::y`